docs/design/copy-info.md

```diff
@@ -0,0 +1,159 @@
+# Copy Info Design
+
+Authors: [Daniel Ploch](mailto:dploch@google.com)
+
+**Summary:** This Document documents an approach to tracking copy information in
+jj repos, in a way that is compatible with both Git's model and with custom
+backends that have more complicated explicit tracking of copy information. This
+design affects the output of diff commands as well as the results of rebasing
+across remote copies.
+
+## Objective
+
+Implement extensible APIs for recording and retrieving copy info for the
+purposes of diffing and rebasing across renames and copies more accurately.
+This should be performant both for Git, which synthesizes copy info on the fly
+between arbitrary trees, and for custom extensions which may explicitly record
+and re-serve copy info over arbitrarily large commit ranges.
+
+The APIs should be defined in a way that makes it easy for custom backends to
+ignore copy info entirely until they are ready to implement it.
+
+### Read API
+
+Copy information will be served by a single new Backend trait method:
+
+```rust
+/// An individual copy source.
+pub struct CopySource {
+    /// The source path a target was copied from.
+    /// It is not requires that the source path is different than the target
+    /// path. A custom backend may choose to represent 'rollbacks' as copies
+    /// from a file unto itself, from a specific prior commit.
+    path: RepoPathBuf,
+    /// The source commit the target was copied from. If not specified, then the
+    /// parent of the target commit is the source commit. Backends may use this field to implement 'integration' logic, where a source may be periodically merged into a target, similar to a branch, but the branching occurs at the file level rather than the repository level. It also follows naturally that any copy source targeted to a specific commit should avoid copy propagation on rebasing, which is desirable for 'fork' style copies.
+    id: Option<CommitId>,
+    // TODO: Record the FileId here?
+}
+
+/// An individual copy event, from file A -> B.
+pub struct CopyRecord {
+    /// The destination of the copy, B.
+    target: RepoPathBuf,
+    /// The CommitId where the copy took place.
+    id: CommitId,
+    /// The source of the copy, A.
+    source: CopySource,
+}
+
+/// Backend options for fetching copy records.
+pub struct CopyRecordOpts {
+    /// If true, follow transitive copy records. That is, if file A is copied to
+    /// B and then to C, a request for copy records of file C should also return
+    /// the copy from A to B.
+    transitive: bool
+    // TODO: Probably something for git similarity detection
+}
+
+pub trait Backend {
+    /// Get all copy records for `paths` in the dag range `roots..heads`.
+    fn get_copy_records(&self, paths: &[RepoPathBuf], heads: &[CommitId], roots: &[CommitId]) -> BackendResult<Box<dyn Iterator<Item = BackendResult<CopyRecord>> + '_>>;
```

Comment:
nit: it feels like `roots` should come before `heads`

By the way, do you think we'll want to impose any requirement about order of copy records or do we leave it up to the backend? But if it's up to the backend, do we actually benefit from it being an iterator (instead of a vector)?

Reply:
Added a comment about ordering